### ISSUE TITLE
Smarter Resources: Search function

### ISSUE THREAD

**Issue by AtieP** [python-discord/site#366]
We need to make sure the content on our site is easy to locate. A search bar would be a welcome way to address this.

## Specifications
- The search bar must be able to search all the content, whether they're static pages, markdown pages or resources.
- The results should be displayed on a separate search results page.
- We should allow the search bar to make **fuzzy matches**. For example, if the user inputs `"aks question"`, the `"asking good questions"` article should be one of the hits.
- The search bar should not be displayed on the homepage, but should be displayed on most other pages.

_edited by lemonsaurus_

**Comment by kwzrd** [comment]
I may be interested in trying this, after I'm done with https://github.com/python-discord/bot/pull/993.

It seems like a nice project that'd allow me to learn more about Django and our site while not being critically important.

**Comment by lemonsaurus** [comment]
This is blocked by the major changes we've got planned for our content management system.

I'm going to add this issue to the Dewikification project, and edit the description with a more precise spec.

**Comment by whispersofthedawn** [comment]
@kwzrd Hey, are you still interested in working on this issue? I'd love to give it a try if you no longer want to implement it.

**Comment by kwzrd** [comment]
@dawnofmidnight You can take it; it's been blocked for a long time and I forgot about it.

**Comment by whispersofthedawn** [comment]
Will do, thank you!

**Comment by whispersofthedawn** [comment]
May I put this one up for grabs? I'm not sure how to implement this, and I don't have enough experience with backend Python.

**Comment by jb3** [comment]
For sure, done!

**Comment by Kronifer** [comment]
could this be done with regexing webpage content?

**Comment by jb3** [comment]
Kind of.

The problem arises that we probably want a site wide search, but we need to render every page before we know what is on there, we can't just search templates and resolve them backwards to the actual page they make up.

It's a tricky one that I'm not sure of the perfect answer for, we need to be able to search the content, resources and events apps, all of which use different rendering methods (only content kind of supports searching since the documents are in their final form).

**Comment by Kronifer** [comment]
i cant think of a way to do this unfortunately.

**Comment by lemonsaurus** [comment]
We can just render all the templates when the site launches, cache the content, the URL and any other metadata somewhere, and then search that metadata?

**Comment by lemonsaurus** [comment]
This is probably blocked by #582, but we should still do it after we're done with that.

**Comment by lemonsaurus** [comment]
The way I see it now, there are really two different search bars we want:
- A reactive search bar on the resources page, which filters down the resources which are shown with immediate effect. Every keypress should do a search.
- A full-site search, for searching the content of pages and articles.

These probably shouldn't be the same search bar, so we should split this into two different issues.

Since this issue is full of kind of confusing commentary that no longer fits, I'll close this issue and open two new ones that better describe what we should be building.

**Comment by lemonsaurus** [comment]
Actually, on second thought, I'm not 100% sure the full-site search fits the app well. It's unclear how static page search would work, and doing only content page search seems kind of unintuitive and like poor UX. 

I'm not sure it's worth implementing, and it would be tricky. Let's start with resource filtering and see if anyone creates a new issue for full-site search.
